hail/src/main/scala/is/hail/nativecode/NativeCodeNode.scala

```diff
@@ -0,0 +1,201 @@
+package is.hail.nativecode
+
+import is.hail.utils.ArrayBuilder
+
+trait NativeCodeUnit {
```

Comment:
This might be pedantic, but I think we should be precious about native code or function pointer (e.g. a pointer to x86 instructions that could be written in many languages) and some CXX that could get compiled and loaded and turned into a function pointer.

I might put these in a separate `cxx` module.  Since you're using the `cxx` module, I wouldn't prefix with Native (or CXX).

I think the tree should be immutable.  There's some naming issues, so I suggest:
 - TranslationUnit(preamble: String, definitions: Array[Definition])
 - Definition (functions and variables)
 - Function(returnType: String, name: String, args: Array[Variable], body: Block) (specifically, a function definition)
 - Statement (this is something that normally ends with a semi-colon)
 - Block(stmts: Array[Statement]) (also known as a compound statement ... slight different than a scope, which includes things like namespaces, class definitions, etc.  Is a CXXStatement.)
  - Expr (a value with a type ... even if we don't track it.  Expr should be implicitly convertible to Statement.)
 - Code (any kind of code, statement or expression)
 - Variable(t: String, name: String, init: String = null) (variable definition)

For simplicity I'd probably make Variable.toString just print out the variable name, just do $v and $v = $e instead of .load and .store (but it will still need a .define() to kick out the definition.)

These should be immutable.  Since you want to build blocks incrementally, I think it's natural to have a:

class BlockBuilder {
  def +=(s: Statement): Unit
  def result(): Block
}

and maybe a TranslationUnitBuilder and maybe if and the loops?